### PR TITLE
[Inductor] Chained sequential autotuning for combo kernels

### PR DESCRIPTION
**Pull request by karthickai**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #177725
* #179041
* __->__ #178936

Replace the flat phase-config approach for combo kernel per-subkernel block tuning with a two-phase chained sequential autotune:

- **Phase 1 (block sizes):** Tune each sub-kernel's block sizes one at a time (largest numel first), with num_warps/num_stages fixed at the base config values. Each step builds on the previous winner so later sub-kernels are benchmarked in the context of earlier winners.

- **Phase 2 (warps/stages):** With block sizes finalized, try all unique (num_warps, num_stages) pairs collected from every heuristic config of every sub-kernel to find the best match for the final block layout.

The old approach benchmarked all phase configs independently against heuristic defaults, missing interactions between sub-kernels. The chained approach captures these interactions at the same O(N*K) benchmark cost.

For given code
```python
import torch
torch._inductor.config.combo_kernels = True
torch._inductor.config.combo_kernel_per_subkernel_blocks = True

def fn(a, b, c, d, e, f):
    return a.cos(), b.sin(), c.exp(), d.neg(), e.relu(), f.sigmoid()

inps = [
    torch.rand(4, 65536, device="cuda"),
    torch.rand(256, 65536, device="cuda"),
    torch.rand(8, 8, device="cuda"),
    torch.rand(1024, 32, device="cuda"),
    torch.rand(16, 4096, device="cuda"),
    torch.rand(128, 128, device="cuda"),
]
fn_c = torch.compile(fn)
out = fn_c(*inps)
```

generated debug log
```bash
  Phase 1 baseline: {'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256} warps=4 time=0.066720
  Phase 1 group 0 SK[1]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] trial={'XBLOCK_0': 1024, 'XBLOCK_1': 512, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256} time=0.066848
  Phase 1 group 0 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 1 group 1 SK[0]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] trial={'XBLOCK_0': 512, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256} time=0.067072
  Phase 1 group 1 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 1 group 2 SK[4]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] trial={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 256, 'XBLOCK_5': 256} time=0.067136
  Phase 1 group 2 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 1 group 3 SK[3]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] trial={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 128, 'XBLOCK_4': 512, 'XBLOCK_5': 256} time=0.067136
  Phase 1 group 3 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 1 group 4 SK[5]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] trial={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 128} time=0.066912
  Phase 1 group 4 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 1 group 5 SK[2]: trying 2 configs, current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
    cfg[0] skip (same as current)
    cfg[1] skip (same as current)
  Phase 1 group 5 winner: current_kwargs={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}
  Phase 2: blocks={'XBLOCK_0': 1024, 'XBLOCK_1': 1024, 'XBLOCK_2': 64, 'XBLOCK_3': 256, 'XBLOCK_4': 512, 'XBLOCK_5': 256}, trying 3 warp/stage pairs
    warps=4 stages=1 skip (same as current)
    warps=8 stages=1 time=0.066912
    warps=1 stages=1 time=0.067232
Combo sequential autotune for triton_poi_fused_0: best config XBLOCK_0: 1024, XBLOCK_1: 1024, XBLOCK_2: 64, XBLOCK_3: 256, XBLOCK_4: 512, XBLOCK_5: 256, num_warps: 4, num_ctas: 1, num_stages: 1, maxnreg: None, time 0.066720
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: torch/_inductor/runtime/triton_heuristics.py [hunk 1]
```diff
@@ -1298,6 +1298,146 @@ def autotune_to_one_config(self, *args, **kwargs):
                 triton_cache_hash=launcher.cache_hash,
             )
 
+    def _combo_sequential_autotune(self, launcher, *args, **kwargs):
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:1301]: can we generalize the multi phase approach, so that it also works with max-autotune ? granted, there may not be many cases where coordinate descent tuning is on but not max autotune.

Edit, jk, this is max-autotune. I think we'll need to generalize something like this for combo kernels as well. (iteratively allow changes of sub configs, then combine).

File: torch/_inductor/runtime/triton_heuristics.py [hunk 2]
```diff
@@ -1298,6 +1298,146 @@ def autotune_to_one_config(self, *args, **kwargs):
...
+        """
+        Chain block-size decisions for combo kernels: tune one group at a time,
+        each step building on the previous winner.
+
+        Phase 1: Tune block sizes with warps/stages fixed from the base config.
+        Phase 2: Re-tune warps/stages with finalized block sizes.
+        """
+        combo_tuning_groups = self.inductor_meta.get("combo_tuning_groups")
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:1309]: Review with help from claude checked out locally let me know your thoughts:

  Bugs                                                                                                                                     
                                                                  
  1. Runs on every call, not just the first (P0)

  triton_heuristics.py:1676 — self.inductor_meta.get("combo_tuning_groups") is always truthy after the first call. Unlike                  
  autotune_to_one_config (gated by len(self.launchers) > 1) or coordesc (gated by found_by_coordesc), there's no "already done" guard.
  Confirmed: 3 calls to the compiled function = 3 full autotune runs with benchmarking. Second call takes ~0.35s vs ~0.05s for subsequent  
  cached calls.                                                   

  2. No cache hook called

  autotune_to_one_config calls self.save_cache_hook to persist the winner. _combo_sequential_autotune doesn't, so the sequentially-tuned   
  config is never cached. Recompilation will re-tune from scratch.
                                                                                                                                           
  Pre-existing issue (not introduced by this PR)                  

  Coordesc is blind to suffixed block-size keys. CoordescTuner.tunable_fields looks for XBLOCK, YBLOCK, etc., but combo kernel configs have
   XBLOCK_0, YBLOCK_0. So coordesc only tunes num_warps for combo kernels — no independent per-axis block refinement. This was already the
  case in the parent commit. Worth a follow-up but not a blocker for this PR.                                                              
                                                                  
  Minor issues

  - No dynamo_timed wrapping — compile-time spent in sequential autotuning is invisible in telemetry breakdowns.                           
  - No counter increments — unlike coordesc which tracks counters["inductor"]["coordesc_tuning_bench"], benchmark calls here aren't
  counted.                                                                                                                                 
  - combo_warp_stage_candidates not guarded — self.inductor_meta.get("combo_warp_stage_candidates") at line 1380 is iterated without a
  fallback. Both keys are always set together so unlikely to crash, but a get(..., []) would be cheap insurance.                           
  - member_indices is always a single-element list — the field name and iteration suggest future grouping support. If planned, a comment
  would help; if not, simplify to a scalar.                                                                                                
                                                                  
  Tests                                                                                                                                    
                                                                  
  - Existing tests updated correctly to check for the new log message.
  - test_combo_autotune_many_subkernels with 6 sub-kernels is good coverage.
  - Tests verify correctness and that the codepath is hit, which is appropriate.

**Review by eellison — COMMENTED** [review]
looks good - a few comments

File: torch/_inductor/runtime/triton_heuristics.py (hunk 2 above)
Comment (karthickai) [reply torch/_inductor/runtime/triton_heuristics.py:1309]: > Runs on every call, not just the first (P0)

Added `found_by_combo_autotune` flag on the config object (same pattern as `found_by_coordesc`). 

> No cache hook called

Added `self.save_cache_hook(launcher.config, self.autotune_time_taken_ns)` at the end of `_combo_sequential_autotune`. 

> Pre-existing issue (not introduced by this PR) Coordesc

Already fixed in the https://github.com/pytorch/pytorch/pull/177725 PR


1) Wrapped the call site with `dynamo_timed("CachingAutotuner.combo_sequential_autotune", ...)`, matching the coordesc pattern.

2) Added `counters["inductor"]["combo_autotune_bench"] += 1` after each of the 3 benchmark call sites (baseline, Phase 1 trials, Phase 2 trials).

3) Not adding a fallback. Both `combo_tuning_groups` and `combo_warp_stage_candidates` are always set together in `_handle_combo_kernel_per_subkernel_blocks`. Adding `get(..., [])` would silently mask a real bug if that invariant ever broke.

4) Grouping is the next PR in the stack (https://github.com/pytorch/pytorch/pull/179041). Sub-kernels with identical heuristic configs are grouped into a single tuning group (e.g. `member_indices: [0, 2]`) to skip redundant benchmark rounds. The existing iteration over `member_indices` already handles this correctly.

**Review by eellison — APPROVED** [review]
